lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -1154,8 +1212,31 @@ estimateKeyWitnessCount utxo txbody@(Cardano.TxBody txbodycontent) =
 
     dummyKeyRole = Payment
 
-    isStakeKeyRegCert (Cardano.StakeAddressRegistrationCertificate _) = True
-    isStakeKeyRegCert _ = False
+    apiScriptHashes =
+        let walletScripts = mapMaybe toTimelockScript scripts
+            toScriptHash =
+                Cardano.hashScript
+                . Cardano.SimpleScript
+                . toCardanoSimpleScript
+            toPair s = (s, toScriptHash s)
+        in toPair <$> walletScripts
+
+    estimateWitNumForCred = \case
+        Cardano.StakeCredentialByKey _ -> 1
+        Cardano.StakeCredentialByScript scriptHash ->
+            let pair = filter (\(_,sh) -> sh == scriptHash) apiScriptHashes
+            in case pair of
+                [(delScript,_)] ->
+                    estimateMaxWitnessRequiredPerInput delScript
+                _ -> error "there should be delegation script in the tx body"
+
+    estimateDelegSigningKeys = \case
+        Cardano.StakeAddressRegistrationCertificate _ -> 0
+        Cardano.StakeAddressDeregistrationCertificate cred ->
+            estimateWitNumForCred cred
+        Cardano.StakeAddressDelegationCertificate cred _ ->
+            estimateWitNumForCred cred
+        _ -> 1
```

Comment:
Ah perfect, nice 👍

Reply:
Or... no wait, the script credential case should be redundant with

```haskell
scriptVkWitsUpperBound =
             fromIntegral
             $ sumVia estimateMaxWitnessRequiredPerInput
            $ mapMaybe toTimelockScript scripts
```
above.

So I think if we just replace the `estimateWitNumForCred cred` with `0` in `estimateDelegSigningKeys`, it should be better.

Reply:
I tried doing this: [try-fixing-estimateKeyWitCount.patch](https://github.com/input-output-hk/cardano-wallet/files/11204174/try-fixing-estimateKeyWitCount.patch)

 Unfortunately we then run into this error:


```haskell
  integration/src/Test/Integration/Scenario/API/Shared/Transactions.hs:1923:15:
  1) API Specifications, SHARED_TRANSACTIONS, SHARED_TRANSACTIONS_DELEGATION_01a - Can join stakepool, rejoin another and quit
       From the following response: Left
           ( ClientError
               ( Object
                   ( fromList
                       [
                           ( "code"
                           , String "created_invalid_transaction"
                           )
                       ,
                           ( "message"
                           , String "The submitted transaction was rejected by the local node. Here's an error message that may help with debugging: HardForkApplyTxErrFromEra S (S (S (S (S (Z (WrapApplyTxErr {unwrapApplyTxErr = ApplyTxError [UtxowFailure (UtxoFailure (AlonzoInBabbageUtxoPredFailure (FeeTooSmallUTxO (Coin 176900) (Coin 166800))))]})))))) Here is the tx: ShelleyTx ShelleyBasedEraBabbage     ( AlonzoTx         { body = TxBodyConstr TxBodyRaw             { _spendInputs = fromList                 [ TxIn                     ( TxId                         { _unTxId = SafeHash "2a96945fbefe340d7039095914ad82a805e9470dd2c9efbc900a89d0f40d2989" }                     )                     ( TxIx 0 )                 ]             , _collateralInputs = fromList []             , _referenceInputs = fromList []             , _outputs = StrictSeq                 { fromStrict = fromList                     [ Sized                         { sizedValue =                             ( Addr Mainnet                                 ( ScriptHashObj                                     ( ScriptHash "0270c7a715297905cac542c23c579bc126427e9fba734dddf6f63cdd" )                                 )                                 ( StakeRefBase                                     ( ScriptHashObj                                         ( ScriptHash "ff1086611634f68a35297e4d136400e18bd42f8ce3af4e337396a9bf" )                                     )                                 )                             , MaryValue 99998833200                                 ( fromList [] )                             , NoDatum                             , SNothing                             )                         , sizedSize = 71                         }                     ]                 }             , _collateralReturn = SNothing             , _totalCollateral = SNothing             , _certs = StrictSeq                 { fromStrict = fromList                     [ DCertDeleg                         ( RegKey                             ( ScriptHashObj                                 ( ScriptHash "ff1086611634f68a35297e4d136400e18bd42f8ce3af4e337396a9bf" )                             )                         )                     , DCertDeleg                         ( Delegate                             ( Delegation                                 { _delegator = ScriptHashObj                                     ( ScriptHash "ff1086611634f68a35297e4d136400e18bd42f8ce3af4e337396a9bf" )                                 , _delegatee = KeyHash "1b3dc19c6ab89eaffc8501f375bb03c11bf8ed5d183736b1d80413d6"                                 }                             )                         )                     ]                 }             , _wdrls = Wdrl                 { unWdrl = fromList [] }             , _txfee = Coin 166800             , _vldt = ValidityInterval                 { invalidBefore = SJust                     ( SlotNo 0 )                 , invalidHereafter = SJust                     ( SlotNo 36060 )                 }             , _update = SNothing             , _reqSignerHashes = fromList []             , _mint = MaryValue 0                 ( fromList [] )             , _scriptIntegrityHash = SNothing             , _adHash = SNothing             , _txnetworkid = SNothing             }         , wits = TxWitnessRaw             { _txwitsVKey = fromList                 [ WitVKeyInternal                     { wvkKey = VKey                         ( VerKeyEd25519DSIGN "1cf5424743e3a0336f230cd276be03a3de9200bdb273b722022fd193f1e9895d" )                     , wvkSig = SignedDSIGN                         ( SigEd25519DSIGN "273bfd43373979413453a9dfca0a8d0c2f634249a21291067af9e67603c25b860cc00ab455dc34433d8ee26a520ec25713d69d9fd218eaae1fc34c1bd7a5440f" )                     , wvkKeyHash = KeyHash "2f9ff4f1f66badab65790cb682c883db544bb738b0f60f0a91d6a715"                     , wvkBytes = "\x82X \x1cõBGCã 3o#\xcÒv¾\x3£Þ\x92\x0½²s·"\x2/Ñ\x93ñé\x89]X@';ýC79yA4S©ßÊ                       \x8d\xc/cBI¢\x12\x91\x6zùæv\x3Â[\x86\xcÀ                       ´UÜ4C=\x8eâjR\xeÂW\x13Ö\x9d\x9fÒ\x18ê®\x1fÃL\x1b×¥D\xf"                     }                 , WitVKeyInternal                     { wvkKey = VKey                         ( VerKeyEd25519DSIGN "68e6d1a3c63cbd8db90904d37282517d989784ddbec41b9147ef7f53fff3fa0f" )                     , wvkSig = SignedDSIGN                         ( SigEd25519DSIGN "859290a3661d213ab3637ec906e8664ff23be14df0add71ebca68330f5a2beb993a97c23227cf1878d5cbca271d314e46d11a84212887f09257abb66b7e9cd0d" )                     , wvkKeyHash = KeyHash "48073afca7453a4862c76fe52c06510a26a4c29fb216efac6c1e1606"                     , wvkBytes = "\x82X hæÑ£Æ<½\x8d¹\x9\x4Ór\x82Q}\x98\x97\x84Ý¾Ä\x1b\x91Gï\x7fSÿóú\xfX@\x85\x92\x90£f\x1d!:³c~É\x6èfOò;áMð\xad×\x1e¼¦\x830õ¢¾¹\x93©|#"|ñ\x87\x8d\¼¢qÓ\x14äm\x11¨B\x12\x88\x7f\x9%z»f·éÍ\xd"                     }                 , WitVKeyInternal                     { wvkKey = VKey                         ( VerKeyEd25519DSIGN "cbaca08d7c8f8e62ec718d71f75626f253eb4932870fc66619433fa99265f0f9" )                     , wvkSig = SignedDSIGN                         ( SigEd25519DSIGN "31b05cc31d43b53d0a1981bd08b6a5da2a3b2b212b8b82305705de2b3a6b30d63ab0aef95927cb7b3b65e95ce19110b61d5d0dca5069d48e10fa117764a3a400" )                     , wvkKeyHash = KeyHash "62d7f6c4fd8b483841fa344008bad7c7c835eb9559c97b5c4ed91ce2"                     , wvkBytes = "\x82X Ë¬ \x8d|\x8f\x8ebìq\x8dq÷V&òSëI2\x87\xfÆf\x19C?©\x92eðùX@1°\Ã\x1dCµ=                       \x19\x81½\x8¶¥Ú*;+!+\x8b\x820W\x5Þ+:k0Ö:°®ùY'Ë{;eé\á\x91\x10¶\x1d]\xdÊPiÔ\x8e\x10ú\x11wd£¤\x0"                     }                 , WitVKeyInternal                     { wvkKey = VKey                         ( VerKeyEd25519DSIGN "b4e9a2469d41a6cf4bd70ab62c4b5f99b4512eec5cc9560d005e9f154ede0336" )                     , wvkSig = SignedDSIGN                         ( SigEd25519DSIGN "226e763ae879ed25b712056c025ad8dd22b8476939d2516a7e7c6ab981ab63b7857f95d7b80fb2a71efc35377cf8981db4ed6bea956d6f1e1697f76c46e81a03" )                     , wvkKeyHash = KeyHash "dda549acce7d5301739214cd871f15dd6efb37dc19fdb18fa723f181"                     , wvkBytes = "\x82X ´é¢F\x9dA¦ÏK×                       ¶,K_\x99´Q.ì\ÉV\xd\x0^\x9f\x15NÞ\x36X@"nv:èyí%·\x12\x5l\x2ZØÝ"¸Gi9ÒQj~|j¹\x81«c·\x85\x7f\x95×¸\xf²§\x1eü57|ø\x98\x1d´íkê\x95mo\x1e\x16\x97÷lFè\x1a\x3"                     }                 ]             , _txwitsBoot = fromList []             , _txscripts = fromList                 [                     ( ScriptHash "e1b97e195443fb0b73ffd08ed0ec05de957e61a0a8af7004cff324c0"                     , TimelockScript TimelockConstr AnyOf                         ( StrictSeq                             { fromStrict = fromList                                 [ TimelockConstr Signature                                     ( KeyHash "62d7f6c4fd8b483841fa344008bad7c7c835eb9559c97b5c4ed91ce2" )                                 , TimelockConstr Signature                                     ( KeyHash "dda549acce7d5301739214cd871f15dd6efb37dc19fdb18fa723f181" )                                 ]                             }                         )                     )                 ,                     ( ScriptHash "ff1086611634f68a35297e4d136400e18bd42f8ce3af4e337396a9bf"                     , TimelockScript TimelockConstr AllOf                         ( StrictSeq                             { fromStrict = fromList                                 [ TimelockConstr Signature                                     ( KeyHash "2f9ff4f1f66badab65790cb682c883db544bb738b0f60f0a91d6a715" )                                 , TimelockConstr Signature                                     ( KeyHash "48073afca7453a4862c76fe52c06510a26a4c29fb216efac6c1e1606" )                                 ]                             }                         )                     )                 ]             , _txdats = TxDatsRaw                 ( fromList [] )             , _txrdmrs = RedeemersRaw                 ( fromList [] )             }         , isValid = IsValid True         , auxiliaryData = SNothing         }     )"
                           )
                       ]
                   )
               )
           )
       While verifying value:
         ( Status
             { statusCode = 500
             , statusMessage = "Internal Server Error"
             }
         , Left
             ( ClientError
                 ( Object
                     ( fromList
                         [
                             ( "code"
                             , String "created_invalid_transaction"
                             )
                         ,
                             ( "message"
                             , String "The submitted transaction was rejected by the local node. Here's an error message that may help with debugging: HardForkApplyTxErrFromEra S (S (S (S (S (Z (WrapApplyTxErr {unwrapApplyTxErr = ApplyTxError [UtxowFailure (UtxoFailure (AlonzoInBabbageUtxoPredFailure (FeeTooSmallUTxO (Coin 176900) (Coin 166800))))]})))))) Here is the tx: ShelleyTx ShelleyBasedEraBabbage     ( AlonzoTx         { body = TxBodyConstr TxBodyRaw             { _spendInputs = fromList                 [ TxIn                     ( TxId                         { _unTxId = SafeHash "2a96945fbefe340d7039095914ad82a805e9470dd2c9efbc900a89d0f40d2989" }                     )                     ( TxIx 0 )                 ]             , _collateralInputs = fromList []             , _referenceInputs = fromList []             , _outputs = StrictSeq                 { fromStrict = fromList                     [ Sized                         { sizedValue =                             ( Addr Mainnet                                 ( ScriptHashObj                                     ( ScriptHash "0270c7a715297905cac542c23c579bc126427e9fba734dddf6f63cdd" )                                 )                                 ( StakeRefBase                                     ( ScriptHashObj                                         ( ScriptHash "ff1086611634f68a35297e4d136400e18bd42f8ce3af4e337396a9bf" )                                     )                                 )                             , MaryValue 99998833200                                 ( fromList [] )                             , NoDatum                             , SNothing                             )                         , sizedSize = 71                         }                     ]                 }             , _collateralReturn = SNothing             , _totalCollateral = SNothing             , _certs = StrictSeq
          { fromStrict = fromList                     [ DCertDeleg                         ( RegKey                             ( ScriptHashObj                                 ( ScriptHash "ff1086611634f68a35297e4d136400e18bd42f8ce3af4e337396a9bf" )                             )                         )                     , DCertDeleg                         ( Delegate                             ( Delegation                                 { _delegator = ScriptHashObj                                     ( ScriptHash "ff1086611634f68a35297e4d136400e18bd42f8ce3af4e337396a9bf" )                                 , _delegatee = KeyHash "
1b3dc19c6ab89eaffc8501f375bb03c11bf8ed5d183736b1d80413d6"                                 }                             )                         )                     ]                 }             , _wdrls = Wdrl                 { unWdrl = fromList [] }             , _txfee = Coin 166800             , _vldt = ValidityInterval                 { invalidBefore = SJust                     ( SlotNo 0 )                 , invalidHereafter = SJust                     ( SlotNo 36060 )                 }             , _update = SNothing             , _reqSignerHashes = fromList []             , _mint = MaryValue 0                 ( fromList [
] )             , _scriptIntegrityHash = SNothing             , _adHash = SNothing             , _txnetworkid = SNothing             }         , wits = TxWitnessRaw             { _txwitsVKey = fromList                 [ WitVKeyInternal                     { wvkKey = VKey                         ( VerKeyEd25519DSIGN "1cf5424743e3a0336f230cd276be03a3de9200bdb273b722022fd193f1e9895d" )                     , wvkSig = SignedDSIGN                         ( SigEd25519DSIGN "273bfd43373979413453a9dfca0a8d0c2f634249a21291067af9e67603c25b860cc00ab455dc34433d8ee26a520ec25713d69d9fd218eaae1fc34c1bd7a5440f" )                     , wvkKeyHash = Key
Hash "2f9ff4f1f66badab65790cb682c883db544bb738b0f60f0a91d6a715"                     , wvkBytes = "\x82X \x1cõBGCã 3o#\xcÒv¾\x3£Þ\x92\x0½²s·"\x2/Ñ\x93ñé\x89]X@';ýC79yA4S©ßÊ                       \x8d\xc/cBI¢\x12\x91\x6zùæv\x3Â[\x86\xcÀ                       ´UÜ4C=\x8eâjR\xeÂW\x13Ö\x9d\x9fÒ\x18ê®\x1fÃL\x1b×¥D\xf"                     }                 , WitVKeyInternal                     { wvkKey = VKey                         ( VerKeyEd25519DSIGN "68e6d1a3c63cbd8db90904d37282517d989784ddbec41b9147ef7f53fff3fa0f" )                     , wvkSig = SignedDSIGN                         ( SigEd25519DSIGN "859290a3661d213ab3637ec906e8664ff23be
14df0add71ebca68330f5a2beb993a97c23227cf1878d5cbca271d314e46d11a84212887f09257abb66b7e9cd0d" )                     , wvkKeyHash = KeyHash "48073afca7453a4862c76fe52c06510a26a4c29fb216efac6c1e1606"                     , wvkBytes = "\x82X hæÑ£Æ<½\x8d¹\x9\x4Ór\x82Q}\x98\x97\x84Ý¾Ä\x1b\x91Gï\x7fSÿóú\xfX@\x85\x92\x90£f\x1d!:³c~É\x6èfOò;áMð\xad×\x1e¼¦\x830õ¢¾¹\x93©|#"|ñ\x87\x8d\¼¢qÓ\x14äm\x11¨B\x12\x88\x7f\x9%z»f·éÍ\xd"                     }                 , WitVKeyInternal                     { wvkKey = VKey                         ( VerKeyEd25519DSIGN "cbaca08d7c8f8e62ec718d71f75626f253eb4932870fc66619433fa99265f0f9" )
  , wvkSig = SignedDSIGN                         ( SigEd25519DSIGN "31b05cc31d43b53d0a1981bd08b6a5da2a3b2b212b8b82305705de2b3a6b30d63ab0aef95927cb7b3b65e95ce19110b61d5d0dca5069d48e10fa117764a3a400" )                     , wvkKeyHash = KeyHash "62d7f6c4fd8b483841fa344008bad7c7c835eb9559c97b5c4ed91ce2"                     , wvkBytes = "\x82X Ë¬ \x8d|\x8f\x8ebìq\x8dq÷V&òSëI2\x87\xfÆf\x19C?©\x92eðùX@1°\Ã\x1dCµ=                       \x19\x81½\x8¶¥Ú*;+!+\x8b\x820W\x5Þ+:k0Ö:°®ùY'Ë{;eé\á\x91\x10¶\x1d]\xdÊPiÔ\x8e\x10ú\x11wd£¤\x0"                     }                 , WitVKeyInternal                     { wvkKey = VKey
  ( VerKeyEd25519DSIGN "b4e9a2469d41a6cf4bd70ab62c4b5f99b4512eec5cc9560d005e9f154ede0336" )                     , wvkSig = SignedDSIGN                         ( SigEd25519DSIGN "226e763ae879ed25b712056c025ad8dd22b8476939d2516a7e7c6ab981ab63b7857f95d7b80fb2a71efc35377cf8981db4ed6bea956d6f1e1697f76c46e81a03" )                     , wvkKeyHash = KeyHash "dda549acce7d5301739214cd871f15dd6efb37dc19fdb18fa723f181"                     , wvkBytes = "\x82X ´é¢F\x9dA¦ÏK×                       ¶,K_\x99´Q.ì\ÉV\xd\x0^\x9f\x15NÞ\x36X@"nv:èyí%·\x12\x5l\x2ZØÝ"¸Gi9ÒQj~|j¹\x81«c·\x85\x7f\x95×¸\xf²§\x1eü57|ø\x98\x1d´íkê\x95mo\x1e\x16\x97÷lFè\x1a\x3"                     }                 ]             , _txwitsBoot = fromList []             , _txscripts = fromList                 [                     ( ScriptHash "e1b97e195443fb0b73ffd08ed0ec05de957e61a0a8af7004cff324c0"                     , TimelockScript TimelockConstr AnyOf                         ( StrictSeq                             { fromStrict = fromList                                 [ TimelockConstr Signature                                     ( KeyHash "62d7f6c4fd8b483841fa344008bad7c7c835eb9559c97b5c4ed91ce2" )                                 , TimelockConstr Signature                                     ( KeyHash "dda549acce7d5301739214cd871f15dd6efb37dc19fdb18fa723f181" )                                 ]                             }                         )                     )                 ,                     ( ScriptHash "ff1086611634f68a35297e4d136400e18bd42f8ce3af4e337396a9bf"                     , TimelockScript TimelockConstr AllOf                         ( StrictSeq                             { fromStrict = fromList                                 [ TimelockConstr Signature                                     ( KeyHash "2f9ff4f1f66badab65790cb682c883db544bb738b0f60f0a91d6a715" )                                 , TimelockConstr Signature                                     ( KeyHash "48073afca7453a4862c76fe52c06510a26a4c29fb216efac6c1e1606" )                                 ]                             }                         )                     )                 ]             , _txdats = TxDatsRaw                 ( fromList [] )             , _txrdmrs = RedeemersRaw                 ( fromList [] )             }         , isValid = IsValid True         , auxiliaryData = SNothing         }     )"
                             )
                         ]
                     )
                 )
             )
         )
       expected: Status {
                   statusCode = 202,
                   statusMessage = "Accepted"
                 }
        but got: Status {
                   statusCode = 500,
                   statusMessage = "Internal Server Error"
                 }

```

Reply:
We have `delegation_template = All [a, b]`, `payment_template = Any [a, b]`, and an estimated key witness count of 3... but 4 witnesses added!

I think in this case the 3 is correct (or at least using "optimal" reasoning).

It then sounds as if `signTx` for the second wallet doesn't realise the input doesn't need to be signed more, and adds a witness anyway... but shouldn't this kind of failure have appeared before in that case? 🤔

Reply:
yes, in the case we adopted for integration testing, we need (party1 and party2) wits from payment template and only one wit from delegation template. It is checked in integration testing when inspecting WitnessCount. Although superfluous the current solution works ...

Reply:
> Although superfluous the current solution works ...

Yes, the test does pass as it is. But if this is because the 1 wit/cert overestimation makes up for the difference between `estimateMaxWitnessRequiredPerInput` and signing all key hashes in the script, then it wouldn't work for bigger script templates — e.g. `Any [a, b, c]`? (could be confirmed)

Possibly a better workaround would be to make the `Any` case here the same as the `All` case
https://github.com/input-output-hk/cardano-wallet/blob/206371ed53a2a9e1e64ba1d84624169431417c5b/lib/wallet/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs#L786
with a comment that it can be changed back, but only if we make signTx smarter